### ISSUE TITLE
Need help with runtime bug (window closes before I want it to)

### ISSUE THREAD

**Issue by awsdert** [stetre/moonnuklear#6]
First I'll start by linking the commit for my project (using specific commit for sake of people who look at this thread after it eventually gets a new commit)
https://github.com/awsdert/gasp/tree/6805541cf5b5adf56b0e531d5e8eefe972fc2438
Now to the problem, after resolving a bug I had with changing the font size (some did not have the correct size) I moved all that code to another file (lua/cfg/font.lua) and modified original code to take the font returned by that dialog (or at least calculate element sizes based on that) and continue as normal, problem is I'm somehow not getting the window to stay open, mind taking a look (at lua/gasp.lua) and seeing if you can identify what I did wrong please?

**Comment by stetre** [comment]
You forgot to render and swap buffers in the rendering loop.

**Comment by awsdert** [comment]
Ah that sorted it, thank you :) Now I just gotta resolve a double initiation issue which should be easier to find and fix

**Comment by awsdert** [comment]
Erm you wouldn't happen to know what might cause this?
> gasp/lua/moonnuklear/glbackend.lua:319: unexpected error (object.c, 100)

**Comment by stetre** [comment]
Might be caused by a double `init()` of the backend without a `shutdown()` in between.

**Comment by awsdert** [comment]
Just tried removing the gl.init() and rear.shutdown() calls and neither removal fixed it, was there a call for initiation in glbackend (rear)? I've also encased the atlas creation into an if statement checking for already initialised fonts and still not fixed :| Gonna look for more initiation code while I wait for a response

**Comment by awsdert** [comment]
Nope can't find any more *.init() calls in font.lua and there's only 2 in the initial gasp.lua, neither of which cause a problem when I create the 1st window, ah btw I'm trying to give the font change it's own window, I've taken note of the glfw.make_context_current() call so I don't think that's it, er just in case you didn't already know it the bug occurs when the font window calls atlas:add()

**Edit:** gasp.lua also calls atlas:add() but only once, all calls thereafter are made in a separate atlas object in font.lua

**Comment by stetre** [comment]
The problem occurs in `font_stash_end()` in the backend, at the creation of the OpenGL texture object for the fonts (line 319 of glbackend.lua, as the error message says).

I'll have a look at it tomorrow morning (now it's late at night in my time zone).

**Comment by awsdert** [comment]
fair enough,' bout time I went to bed too :)

**Comment by stetre** [comment]
The problem arises because you are creating multiple GL windows (with `glfw.create_window()`), and the OpenGL driver reuses the same texture identifier in different windows, which MoonGL doesn't expect.

This is a design bug in MoonGL, I suppose, and I expect the same problem to happen also for GL objects of other kinds (unfortunately the venerable OpenGL uses indices 1, 2, 3 etc. as identifiers...).
I'll investigate to see how to fix this.

Note however, that one GL window is usually enough for an application. The GL window is not to be confused with a _window widget_: it is the area on the screen where the whole application is drawn. Within that area, you can draw as many window widgets you want, even dynamically (which is what you are attempting to do, I presume). If you look at the examples - e.g. `demo.lua` - they are fairly complex with many windows and widgets of all sorts, but any of them uses only one GL window.

**Comment by awsdert** [comment]
Ah so that's why, alright I'll take a look at that demo.lua for now but if you can fix that bug please do, I kinda need to use multiple windows specifically for handling the case where a user wants multiple things to be shown yet not expand the window for it, I did that sorta thing regularly with Cheat Engine (Windows only) and would like similar flexibility with the app I'm developing to stand-in for it until PINCE is functioning properly, for example I liked to keep the memory editor open so I could watch for changes in specific areas while I ran scripts from the main window while also doing stuff in-game in the background (assuming the game supported being played in the background)

**Comment by stetre** [comment]
I'll work on it asap. If you are not familiar with OpenGL programming, though, you'll nonetheless have a hard time using multiple concurrent GL contexts in the same application (and learning OpenGL is not a matter of a few afternoons). It's not that easy as just creating multiple windows: you need also to manually swap between contexts (i.e. make the correct one current) throughout the application, and I'm not sure that the glbackend.lua will work correctly in such a use case (I wrote it assuming 1 context), so you'd probably have to roll your own backend.

On the other hand, Nuklear's window and group widgets have several properties that may suit your needs (see panelflags in the manual, for example). You can layout them in the GL window as you like, make them closable, minimizable, etc.

**Comment by awsdert** [comment]
Since I can't get multiple windows for now I've done a separate method for now, gasp can draw either frame for now and hopefully can seamlessly draw in another window instead when it is supported, anyways if you want to add/change the link (wasn't paying that close an attention to updates) here's the one involving both font change and volume change (well the value to use for it anyways)
https://github.com/awsdert/gasp/tree/b4285dd6acd1b28b046f52b4ab1b0d2b0646a739
If you think anything could be made simpler to understand please do mention it (even as an issue if you so like) :)

**Comment by awsdert** [comment]
Did a restart so I'm gonna close this issue as it is no longer relevant
